### PR TITLE
Creating run and discovery start events and exposing it to loggers

### PR DESCRIPTION
**Pull request by abhishekkumawat23**
**Changes contains**:
1. Creating run and discovery start events
2. Exposing start events to loggers
3. _Exposing test case filters to logger as an discovery event arg._

### CODE REVIEW

File: src/Microsoft.TestPlatform.Client/Discovery/DiscoveryRequest.cs [hunk 1]
```diff
@@ -142,6 +147,21 @@ bool IRequest.WaitForCompletion(int timeout)
...
 
         /// <summary>
+        /// Gets TestCaseFilterExpression from discovery criteria's test case filter
+        /// </summary>
+        private TestCaseFilterExpression GetTestCaseFilter()
+        {
+            FilterExpressionWrapper filterExpressionWrapper = !string.IsNullOrEmpty(this.DiscoveryCriteria.TestCaseFilter) ? new FilterExpressionWrapper(this.DiscoveryCriteria.TestCaseFilter) : null;
+            return (filterExpressionWrapper != null && string.IsNullOrEmpty(filterExpressionWrapper.ParseError)) ?
+                new TestCaseFilterExpression(filterExpressionWrapper) : null;
+        }
```
Comment (navin22) [inline src/Microsoft.TestPlatform.Client/Discovery/DiscoveryRequest.cs]: Can't we do this in the logger end itself. As the entire discovery criteria is already passed as part of the DiscoveryStartEventArgs ?

Comment (abhishekkumawat23) [reply src/Microsoft.TestPlatform.Client/Discovery/DiscoveryRequest.cs]: FilterExpressionWrapper and TestCaseFilterExpression is not exposed to logger. Only ITestCaseFilterExpression interface is exposed and thus cant be done at logger side

File: src/Microsoft.TestPlatform.Client/Discovery/DiscoveryRequest.cs [hunk 2]
```diff
@@ -62,6 +63,10 @@ void IDiscoveryRequest.DiscoverAsync()
                     // Collecting Data Point Number of sources sent for discovery
                     this.requestData.MetricsCollection.Add(TelemetryDataConstants.NumberOfSourcesSentForDiscovery, (this.DiscoveryCriteria.Sources.Count()).ToString());
 
+                    // Invoke OnDiscoveryStart event
+                    TestCaseFilterExpression filterExpression = this.GetTestCaseFilter();
```
Comment (navin22) [inline src/Microsoft.TestPlatform.Client/Discovery/DiscoveryRequest.cs]: >TestCaseFilterExpression filterExpression = this.GetTestCaseFilter(); [](start = 20, length = 69)

Hey if no loggers are registered, then don't make this call.

**Review by navin22 — COMMENTED** [review]
Looks good with suggestions.

**Review by singhsarab — APPROVED** [review]
:shipit:

**Review by Faizan2304 — APPROVED** [review]

**Pull request merged**